File: sw/simulator/gaia.ml
(*
 * World environment (time, wind, ...) for multi-AC simulation
 *
 * Copyright (C) 2004 Pascal Brisset, Antoine Drouin
 *
 * This file is part of paparazzi.
 *
 * paparazzi is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * paparazzi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with paparazzi; see the file COPYING.  If not, write to
 * the Free Software Foundation, 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 *)

open Printf
open Latlong

let my_id = "gaia"
let sending_period = 5000 (* ms *)

module Ground_Pprz = PprzLink.Messages(struct let name = "ground" end)

let ivy_bus = ref Defivybus.default_ivy_bus
let time_scale = ref 1.
let wind_speed = ref 0.
let wind_dir = ref 0.
let wind_up = ref 0.
let gps_off = ref false

let parse_args = fun () ->
  let options =
    [ "-b", Arg.String (fun x -> ivy_bus := x), (sprintf "Bus\tDefault is %s" !ivy_bus);
      "-t", Arg.Set_float time_scale, "Set time scale (default: 1.0)";
      "-w", Arg.Set_float wind_speed, "Set wind speed (0-30m/s)";
      "-d", Arg.Set_float wind_dir, "Set wind direction 0-359 deg";
      "-u", Arg.Set_float wind_up, "Set wind updraft (-10 to 10m/s)";
      "-g", Arg.Set gps_off, "Turn off GPS";
    ] in
  Arg.parse (options)
    (fun x -> Printf.fprintf stderr "Warning: Don't do anythig with %s\n" x)
    "Usage: "

let _ =
  parse_args ();
  let icon = GdkPixbuf.from_file Env.icon_file in
  let window = GWindow.window ~icon ~title:"Gaia" () in
  let quit = fun () -> GMain.Main.quit (); exit 0 in
  ignore (window#connect#destroy ~callback:quit);

  let time_scale_adj = GData.adjustment ~page_size:0. ~value:!time_scale ~lower:(0.5) ~upper:10. ~step_incr:0.5 () in
  let wind_dir_adj = GData.adjustment ~page_size:0. ~value:!wind_dir ~lower:(0.) ~upper:359. ~step_incr:1.0 () in
  let wind_speed_adj = GData.adjustment ~page_size:0. ~value:!wind_speed ~lower:(0.) ~upper:30. ~step_incr:0.1 () in
  let wind_up_adj = GData.adjustment ~page_size:0. ~value:!wind_up ~lower:(-10.) ~upper:10. ~step_incr:0.1 () in
  let infrared_contrast_adj = GData.adjustment ~page_size:0. ~value:266. ~lower:(0.) ~upper:1010. ~step_incr:10. () in
  let gps_sa = GButton.toggle_button ~label:"GPS OFF" ~active:!gps_off () in

  let world_values = fun asker_values ->
    let wind_speed = wind_speed_adj#value
    and wind_dir_rad = Latlong.pi /. 2. -. (Deg>>Rad) wind_dir_adj#value in

    let wind_east = -. wind_speed  *. cos wind_dir_rad
    and wind_north = -. wind_speed *. sin wind_dir_rad in
    let wind_up = wind_up_adj#value in

    [ "wind_east", PprzLink.Float wind_east;
      "wind_north", PprzLink.Float wind_north;
      "wind_up", PprzLink.Float wind_up;
      "ir_contrast", PprzLink.Float infrared_contrast_adj#value;
      "time_scale", PprzLink.Float time_scale_adj#value;
      "gps_availability", PprzLink.Int (if gps_sa#active then 0 else 1)
    ] in
  let world_send = fun () ->
    Ground_Pprz.message_send my_id "WORLD_ENV" (world_values []) in

  List.iter
    (fun (a:GData.adjustment) -> ignore (a#connect#value_changed world_send))
    [time_scale_adj; wind_dir_adj; wind_speed_adj; wind_up_adj; infrared_contrast_adj];
  ignore (gps_sa#connect#toggled world_send);

  ignore (Glib.Timeout.add sending_period (fun () -> world_send (); true));

  let vbox = GPack.vbox ~packing:window#add () in

  let hbox = GPack.hbox ~packing:vbox#pack () in
  let _ =  GMisc.label ~text:"Time scale:" ~packing:hbox#pack () in
  let _ts = GEdit.spin_button ~adjustment:time_scale_adj ~packing:hbox#add () in

  let hbox = GPack.hbox ~packing:vbox#pack () in
  ignore (GMisc.label ~text:"Wind dir:" ~packing:hbox#pack ());
  ignore (GRange.scale ~digits:0 `HORIZONTAL ~adjustment:wind_dir_adj ~packing:hbox#add ());

  let hbox = GPack.hbox ~packing:vbox#pack () in
  ignore (GMisc.label ~text:"Wind speed:" ~packing:hbox#pack ());
  ignore (GRange.scale `HORIZONTAL ~adjustment:wind_speed_adj ~packing:hbox#add ());

  let hbox = GPack.hbox ~packing:vbox#pack () in
  ignore (GMisc.label ~text:"Wind up:" ~packing:hbox#pack ());
  ignore (GRange.scale `HORIZONTAL ~adjustment:wind_up_adj ~packing:hbox#add ());

  vbox#pack gps_sa#coerce;

  Ivy.init "Paparazzi gaia" "READY" (fun _ _ -> ());
  Ivy.start !ivy_bus;

  ignore (Ground_Pprz.message_answerer my_id "WORLD_ENV" (fun _ -> world_values));

  window#show ();
  Unix.handle_unix_error GMain.Main.main ()

File: sw/simulator/gaia.py
#!/usr/bin/env python3

# World environment (time, wind, ...) for multi-AC simulation
#
# Copyright (C) 2024 Mael FEURGARD <mael.feurgard@enac.fr>
# 
# This file is part of paparazzi.
# 
# paparazzi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# paparazzi is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with paparazzi.  If not, see <https://www.gnu.org/licenses/>.

import os
import sys
import math

PPRZ_HOME = os.getenv("PAPARAZZI_HOME", os.path.normpath(os.path.join(os.path.dirname(os.path.abspath(__file__)), '../../../')))
sys.path.append(PPRZ_HOME + "/var/lib/python")

from pprzlink.ivy import IvyMessagesInterface
from pprzlink.message import PprzMessage


from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QWidget,QMainWindow,QApplication,\
                            QLabel

from gaia_python.betterHSlider import BetterHSlider
from gaia_python.AngleSelector import AngleSelector

class Ui_Dialog(object):
    def setupUi(self, Dialog):
        Dialog.setObjectName("Gaia")
        Dialog.resize(663, 331)
        
        self.horizontalLayout_3 = QtWidgets.QHBoxLayout(Dialog)
        self.horizontalLayout_3.setObjectName("horizontalLayout_3")
        
        self.main_horizontalLayout = QtWidgets.QHBoxLayout()
        self.main_horizontalLayout.setObjectName("main_horizontalLayout")
        
        self.sliders_frame = QtWidgets.QFrame(Dialog)
        self.sliders_frame.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.sliders_frame.setFrameShadow(QtWidgets.QFrame.Plain)
        self.sliders_frame.setObjectName("sliders_frame")
        
        self.verticalLayout_4 = QtWidgets.QVBoxLayout(self.sliders_frame)
        self.verticalLayout_4.setObjectName("verticalLayout_4")
        
        self.sliders_verticalLayout = QtWidgets.QVBoxLayout()
        self.sliders_verticalLayout.setObjectName("sliders_verticalLayout")
        
        self.time_slider = BetterHSlider(self.sliders_frame)
        self.time_slider.setObjectName("time_slider")
        self.time_slider.setText("Time scale")
        self.time_slider.setDecimals(1)
        self.time_slider.setRange(0.1,10)
        self.time_slider.setValue(1)
        self.sliders_verticalLayout.addWidget(self.time_slider)
        
        self.hline_fst = QtWidgets.QFrame(self.sliders_frame)
        self.hline_fst.setFrameShape(QtWidgets.QFrame.HLine)
        self.hline_fst.setFrameShadow(QtWidgets.QFrame.Plain)
        self.hline_fst.setObjectName("hline_fst")
        self.sliders_verticalLayout.addWidget(self.hline_fst)
        
        self.speed_slider = BetterHSlider(self.sliders_frame)
        self.speed_slider.setObjectName("speed_slider")
        self.speed_slider.setText("Wind speed (m/s)")
        self.speed_slider.setDecimals(1)
        self.speed_slider.setRange(0,30)
        self.speed_slider.setValue(0)
        self.sliders_verticalLayout.addWidget(self.speed_slider)
        
        self.hline_snd = QtWidgets.QFrame(self.sliders_frame)
        self.hline_snd.setFrameShape(QtWidgets.QFrame.HLine)
        self.hline_snd.setFrameShadow(QtWidgets.QFrame.Plain)
        self.hline_snd.setObjectName("hline_snd")
        self.sliders_verticalLayout.addWidget(self.hline_snd)
        
        self.up_slider = BetterHSlider(self.sliders_frame)
        self.up_slider.setObjectName("up_slider")
        self.up_slider.setText("Updraft (m/s)")
        self.up_slider.setDecimals(1)
        self.up_slider.setRange(-10,10)
        self.up_slider.setValue(0)
        self.sliders_verticalLayout.addWidget(self.up_slider)
        
        self.hline_trd = QtWidgets.QFrame(self.sliders_frame)
        self.hline_trd.setFrameShape(QtWidgets.QFrame.HLine)
        self.hline_trd.setFrameShadow(QtWidgets.QFrame.Plain)
        self.hline_trd.setObjectName("hline_trd")
        self.sliders_verticalLayout.addWidget(self.hline_trd)
        
        self.gps_checkBox = QtWidgets.QCheckBox(self.sliders_frame)
        self.gps_checkBox.setObjectName("gps_checkBox")
        self.gps_checkBox.setText("GPS disabled")
        self.gps_checkBox.setChecked(False)
        self.sliders_verticalLayout.addWidget(self.gps_checkBox, 0, QtCore.Qt.AlignHCenter)
        self.verticalLayout_4.addLayout(self.sliders_verticalLayout)
        
        self.main_horizontalLayout.addWidget(self.sliders_frame)
        self.dial_frame = QtWidgets.QFrame(Dialog)
        self.dial_frame.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.dial_frame.setFrameShadow(QtWidgets.QFrame.Raised)
        self.dial_frame.setObjectName("dial_frame")
        
        self.horizontalLayout_2 = QtWidgets.QHBoxLayout(self.dial_frame)
        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
        
        self.wind_dial = AngleSelector(self.dial_frame)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.MinimumExpanding, QtWidgets.QSizePolicy.MinimumExpanding)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.wind_dial.sizePolicy().hasHeightForWidth())
        self.wind_dial.setSizePolicy(sizePolicy)
        self.wind_dial.setLayoutDirection(QtCore.Qt.LeftToRight)
        self.wind_dial.setObjectName("wind_dial")
        self.wind_dial.setText("Wind direction (°)")
        self.horizontalLayout_2.addWidget(self.wind_dial)
        self.main_horizontalLayout.addWidget(self.dial_frame)
        self.horizontalLayout_3.addLayout(self.main_horizontalLayout)

        self.retranslateUi(Dialog)
        QtCore.QMetaObject.connectSlotsByName(Dialog)
        
        

    def retranslateUi(self, Dialog):
        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Gaia", "Gaia"))
        self.gps_checkBox.setText(_translate("Dialog", "GPS disabled"))



class Gaia(QWidget,Ui_Dialog):
    def __init__(self, parent: QWidget | None = ..., flags: QtCore.Qt.WindowFlags | QtCore.Qt.WindowType = QtCore.Qt.WindowType.Widget) -> None:
        super().__init__(parent, flags)
        self.setupUi(self)
    

class GaiaWindow(QMainWindow):
    def __init__(self, parent: QWidget | None = None, flags: QtCore.Qt.WindowFlags | QtCore.Qt.WindowType = QtCore.Qt.WindowType.Window) -> None:
        super().__init__(parent, flags)
        
        self.gaiaWidget = Gaia(self)
        
        self.setObjectName("Gaia dialog")
        self.setWindowTitle("Gaia dialog")
        self.setCentralWidget(self.gaiaWidget)
        
        self.ivy = IvyMessagesInterface("gaia")
        self.timer = QtCore.QTimer(self)
        
        self.timer.timeout.connect(self.__send_status)
        
        self.gaiaWidget.gps_checkBox.stateChanged.connect(self.__send_status)
        
        self.gaiaWidget.time_slider.valueChanged.connect(self.__send_status)
        self.gaiaWidget.speed_slider.valueChanged.connect(self.__send_status)
        self.gaiaWidget.up_slider.valueChanged.connect(self.__send_status)
        self.gaiaWidget.wind_dial.valueChanged.connect(self.__send_status)
        
        self.timer.start(5000)
        
    def setUpdateInterval(self,msec:int):
        self.timer.setInterval(msec)
        
    def updateInterval(self) -> int:
        return self.timer.interval()
        
        
    def __send_status(self):
        rad_wind_dir = self.gaiaWidget.wind_dial.value()*math.pi/180
        wind_speed = self.gaiaWidget.speed_slider.value()
        wind_up = self.gaiaWidget.up_slider.value()
        timescale = self.gaiaWidget.time_slider.value()
        
        msg = PprzMessage("ground","WORLD_ENV")
        msg["wind_east"] = - wind_speed * math.sin(rad_wind_dir)
        msg["wind_north"] = - wind_speed * math.cos(rad_wind_dir)
        msg["wind_up"] = wind_up
        msg["ir_contrast"] = 266
        msg["time_scale"] = timescale
        msg["gps_availability"] = 0 if self.gaiaWidget.gps_checkBox.isChecked() else 1
        
        self.ivy.send(msg,self.ivy.agent_name)
        
        
    def closeEvent(self, e: QtGui.QCloseEvent) -> None:
        self.ivy.shutdown()
        e.accept()
        
if __name__ == "__main__":
    app = QApplication([])
    window = GaiaWindow(flags=QtCore.Qt.WindowType.Window)
    window.show()
    app.exec()

